fix(vulnerable-code): Fix search for Go package vulnerabilities

wkl3nk opened the pull request:
For Go packages, both the namespace and name may contain path segments separated by a "/" character. The purl specification requires these "/" characters to be percent-encoded in the namespace and name components of a purl.
The VulnerableCode bulk-search API is unable to handle these percent-encoded "/" characters, resulting in no vulnerability records being returned.
This bugfix decodes any percent-encoded "/" characters just before making the VulnerableCode query to ensure proper functionality.

Fixes #9298

sschuberth commented:
> Fixes #9298

Not: Missing period at the end of the sentence.

wkl3nk commented on plugins/advisors/vulnerable-code/src/main/kotlin/VulnerableCode.kt:
```diff
@@ -93,23 +93,35 @@ class VulnerableCode(override val descriptor: PluginDescriptor, config: Vulnerab
-        val purls = packages.mapNotNull { pkg -> pkg.purl.ifEmpty { null } }
-        val chunks = purls.chunked(BULK_REQUEST_SIZE)
+        // Workaround for #9298
+        // Create a list of pairs, where the first element is the purl without percent encoding
```
No, that unfortunately won't work, because the ORT purls contain a mixture of "/" and "%2F" like in pkg:golang/github.com%2Fquic-go%2Fquic-go@0.40.0, and once you have replaced all "%2F" with "/" there is no more way back to the "mixed" ORT purl. This was the reason for having this implementation as pairs, where the first element is the modified purl, and the second element the original ORT purl.

sschuberth replied:
> there is no more way back to the "mixed" ORT purl.

Correct, but my point is that my implementation proposal should not require you to go back, as the VC-style-PURL is only used for the lookup in the VC results.

Anyway, we're now having a larger discussion about PURL encoding over [here](https://github.com/package-url/purl-spec/pull/176), and I'm tempted to say we should solve this completely differently: By not encoding `/` that occurs as part of the namespace at all, but only in the name.

wkl3nk commented:
> Could you try if https://github.com/oss-review-toolkit/ort/pull/9304 instead would also fix your issue?

This is about namespace segments?
It probably would, but from what I have seen in ort-result.zip

```
id: "Go::github.com/quic-go/quic-go:0.40.0"
purl: "pkg:golang/github.com%2Fquic-go%2Fquic-go@0.40.0"
```
I ask myself if there is not a general problem in detecting `namespace` and `name` for Go libraries.
Would not the following be more correct:

```
id: "Go:github.com/quic-go:quic-go:0.40.0"
purl: "pkg:golang/github.com/quic-go/quic-go@0.40.0"
```
Because the github.com home of quic-go is https://github.com/quic-go and quic-go is just one of the subprojects there.

sschuberth commented:
Yes, that's exactly what my PR is about. But could be that there's still something wrong in my implementation.

wkl3nk commented:
Possibly I did not know enough when I wroted the posting on Friday. Because I am not a Go programmer, I don't even know if there exists a concept of namespaces. Nevertheless I proposed to use the namespace, based on some directory structure on a github.com account. I don't know if this is valid, or if it is better to say something like "There are no namespaces in Go". Don't know.

sschuberth commented:
> I don't know if this is valid, or if it is better to say something like "There are no namespaces in Go".

You may want to refer to [this comment](https://github.com/package-url/purl-spec/pull/176#issuecomment-2420029239).

sschuberth commented:
I still believe we should solve this differently, in a more generic way. While Go apparently does not have namespaces, the purl standard treats them as if Go had namespaces. I'm currently preparing a fix for that.

sschuberth commented:
In any case, @wkl3nk as mentioned [here](https://github.com/oss-review-toolkit/ort/issues/9298#issuecomment-2419213085) please add a test for this (in a new commit that could easily be cherry-picked) so we can verify the fix also for alternative solutions.

wkl3nk commented:
@sschuberth I added a test cases, hoping that this is what you had in mind. It will start to fail once the slashes in the purl name section are no longer percent-encoded. Once this test fails, this commit can be cherry-picked and removed.

sschuberth commented:
> I added a test cases, hoping that this is what you had in mind.

Hmm, not really. I was hoping for a funTest that queries VC for a Go package which is known to have vulnerabilities. When run against `main`, the test would currently fail as the expected vulnerability is not found. But once there is a fix in place to produce the correct purl, the test would succeed.

sschuberth commented:
> > I added a test cases, hoping that this is what you had in mind.
> 
> Hmm, not really. I was hoping for a funTest that queries VC for a Go package which is known to have vulnerabilities.

Nevermind @wkl3nk, I'll now add such a test myself as I'm close to proposing a more generic fix.

sschuberth commented:
Closing in favor of the more generic solution in https://github.com/oss-review-toolkit/ort/pull/9330.
